### PR TITLE
executor: fix wrong behavior of set charset statement

### PR DESCRIPTION
**Pull request by wjhuang2016**
### What problem does this PR solve?

Problem Summary:

Previously, we take `set charset` as `set names`, and set `charsetCollation` too. However, it's wrong, as it's stated in [here](https://dev.mysql.com/doc/refman/8.0/en/set-character-set.html)  

### What is changed and how it works?

Distinguish `set names` and `set charset` statements and handle them differently.
Related Parser [PR](https://github.com/pingcap/parser/pull/836).

### Related changes

- Need to cherry-pick to the release branch

### Check List 

Tests 

- UnitTest

Side effects

### Release note 
- Fix wrong behavior of set charset statement

### CODE REVIEW

**Comment by tiancaiamao** [comment]
```
[2020-05-06T08:56:12.759Z] ----------------------------------------------------------------------
[2020-05-06T08:56:12.759Z] FAIL: set_test.go:421: testSuite5.TestSetCharset
[2020-05-06T08:56:12.759Z] 
[2020-05-06T08:56:12.759Z] set_test.go:500:
[2020-05-06T08:56:12.759Z]     check(
[2020-05-06T08:56:12.759Z]         "latin1",
[2020-05-06T08:56:12.759Z]         "latin1",
[2020-05-06T08:56:12.759Z]         "latin1",
[2020-05-06T08:56:12.759Z]         "utf8mb4",
[2020-05-06T08:56:12.759Z]         "utf8mb4",
[2020-05-06T08:56:12.759Z]         "utf8",
[2020-05-06T08:56:12.759Z]         "binary",
[2020-05-06T08:56:12.759Z]     )
[2020-05-06T08:56:12.759Z] set_test.go:440:
[2020-05-06T08:56:12.759Z]     c.Assert(sVar, Equals, args[i], Commentf("%d: %s", i, characterSetVariables[i]))
[2020-05-06T08:56:12.759Z] ... obtained string = "utf8mb4"
[2020-05-06T08:56:12.759Z] ... expected string = "latin1"
[2020-05-06T08:56:12.759Z] ... 1: character_set_connection
```

@wjhuang2016

**Comment by wjhuang2016** [comment]
> ```
> [2020-05-06T08:56:12.759Z] ----------------------------------------------------------------------
> [2020-05-06T08:56:12.759Z] FAIL: set_test.go:421: testSuite5.TestSetCharset
> [2020-05-06T08:56:12.759Z] 
> [2020-05-06T08:56:12.759Z] set_test.go:500:
> [2020-05-06T08:56:12.759Z]     check(
> [2020-05-06T08:56:12.759Z]         "latin1",
> [2020-05-06T08:56:12.759Z]         "latin1",
> [2020-05-06T08:56:12.759Z]         "latin1",
> [2020-05-06T08:56:12.759Z]         "utf8mb4",
> [2020-05-06T08:56:12.759Z]         "utf8mb4",
> [2020-05-06T08:56:12.759Z]         "utf8",
> [2020-05-06T08:56:12.759Z]         "binary",
> [2020-05-06T08:56:12.759Z]     )
> [2020-05-06T08:56:12.759Z] set_test.go:440:
> [2020-05-06T08:56:12.759Z]     c.Assert(sVar, Equals, args[i], Commentf("%d: %s", i, characterSetVariables[i]))
> [2020-05-06T08:56:12.759Z] ... obtained string = "utf8mb4"
> [2020-05-06T08:56:12.759Z] ... expected string = "latin1"
> [2020-05-06T08:56:12.759Z] ... 1: character_set_connection
> ```
> 
> @wjhuang2016

Fixed.

File: executor/set.go [hunk 1]
```diff
@@ -248,12 +248,33 @@ func (e *SetExecutor) setCharset(cs, co string) error {
...
+	// Set charset statement
+	for _, v := range variable.SetCharsetVariables {
 		if err = sessionVars.SetSystemVar(v, cs); err != nil {
 			return errors.Trace(err)
 		}
 	}
-	return sessionVars.SetSystemVar(variable.CollationConnection, co)
+	csDb, err := sessionVars.GlobalVarsAccessor.GetGlobalSysVar(variable.CharsetDatabase)
+	if err != nil {
+		return err
```
Comment (bb7133) [inline executor/set.go:267]: Why not using `errors.Trace`?

Comment (AilinKid) [inline executor/set.go:267]: better use the same error trace as below.

Comment (wjhuang2016) [reply executor/set.go:267]: The err return by GetGlobalSysVar already has a stack.

File: go.mod [hunk 2]
```diff
@@ -49,14 +49,15 @@ require (
...
-	go.uber.org/zap v1.14.1
+	go.uber.org/zap v1.15.0
 	golang.org/x/net v0.0.0-20200324143707-d3edc9973b7e
 	golang.org/x/sync v0.0.0-20190911185100-cd5d95a43a6e
 	golang.org/x/sys v0.0.0-20200323222414-85ca7c5b95cd
 	golang.org/x/text v0.3.2
 	golang.org/x/tools v0.0.0-20200325203130-f53864d0dba1
 	google.golang.org/grpc v1.25.1
 	gopkg.in/natefinch/lumberjack.v2 v2.0.0
+	honnef.co/go/tools v0.0.1-2020.1.4 // indirect
```
Comment (AilinKid) [inline go.mod:60]: should clean the attached modification except parser

File: executor/set.go [hunk 3]
```diff
@@ -248,12 +248,33 @@ func (e *SetExecutor) setCharset(cs, co string) error {
...
+	}
+	coDb, err := sessionVars.GlobalVarsAccessor.GetGlobalSysVar(variable.CollationDatabase)
+	if err != nil {
+		return err
+	}
+	err = sessionVars.SetSystemVar(variable.CharacterSetConnection, csDb)
```
Comment (AilinKid) [inline executor/set.go:273]: better add a comment about why here need derive the `coDB,` `csDB` globally and set it in session vars additionally

Comment (wjhuang2016) [reply executor/set.go:273]: Added

File: go.mod (hunk 2 above)
Comment (wjhuang2016) [reply go.mod:60]: `make check` adds it automatedly.

**Pull request merged**